shim/src/evaluators.js

```diff
@@ -108,73 +112,87 @@ function createScopedEvaluatorFactory(unsafeRec, constants) {
+    // 'this' binding.
+    const safeEval = {
+      eval(src) {
+        src = `${src}`;
+        rejectImportExpressions(src);
+        scopeHandler.useUnsafeEvaluator = true;
+        let err;
+        try {
+          // Ensure that "this" resolves to the safe global.
+          return apply(scopedEvaluator, safeGlobal, [src]);
+        } catch (e) {
+          // stash the child-code error in hopes of debugging the internal failure
+          err = e;
+          throw e;
+        } finally {
+          // belt and suspenders: the proxy switches this off immediately after
+          // the first access, but just in case we clear it here too
+          if (scopeHandler.useUnsafeEvaluator !== false) {
+            scopeHandler.useUnsafeEvaluator = false;
+            throwTantrum('handler sets useUnsafeEvaluator = false', err);
+          }
         }
       }
-    }
-  }.eval;
+    }.eval;
 
-  // safeEval's prototype is currently the primal realm's Function.prototype,
-  // which we must not let escape. To make 'eval instanceof Function' be true
-  // inside the realm, we need to point it at the RootRealm's value.
+    // safeEval's prototype is currently the primal realm's
+    // Function.prototype, which we must not let escape. To make 'eval
+    // instanceof Function' be true inside the realm, we need to point it at
+    // the RootRealm's value.
 
-  // Ensure that eval from any compartment in a root realm is an
-  // instance of Function in any compartment of the same root realm.
-  setPrototypeOf(safeEval, unsafeFunction.prototype);
+    // Ensure that eval from any compartment in a root realm is an instance
+    // of Function in any compartment of the same root realm.
+    setPrototypeOf(safeEval, unsafeFunction.prototype);
 
-  assert(getPrototypeOf(safeEval).constructor !== Function, 'hide Function');
-  assert(getPrototypeOf(safeEval).constructor !== unsafeFunction, 'hide unsafeFunction');
+    assert(getPrototypeOf(safeEval).constructor !== Function, 'hide Function');
+    assert(getPrototypeOf(safeEval).constructor !== unsafeFunction, 'hide unsafeFunction');
 
-  // note: be careful to not leak our primal Function.prototype by setting
-  // this to a plain arrow function. Now that we have safeEval, use it.
-  defineProperty(safeEval, 'toString', {
-    value: safeEval("() => 'function eval() { [shim code] }'"),
-    writable: false,
-    enumerable: false,
-    configurable: true
-  });
+    // note: be careful to not leak our primal Function.prototype by setting
+    // this to a plain arrow function. Now that we have safeEval, use it.
+    defineProperty(safeEval, 'toString', {
+      value: safeEval("() => 'function eval() { [shim code] }'"),
+      writable: false,
+      enumerable: false,
+      configurable: true
+    });
+
+    return safeEval;
+  }
+
+  return factory;
+}
+
+export function createSafeEvaluator(safeEvaluatorFactory) {
+  return safeEvaluatorFactory({});
```

Comment:
Maybe use `safeEvaluatorFactory()`

Reply:
I'll do that if you like, but in general I'm a fan of fixed arguments rather than optional ones. OTOH a lot of my experience is coming from Python, so if optional args is more idiomatic for JS, I'm all for it.